core/src/oauth/connection.rs

```diff
@@ -338,17 +371,62 @@ impl Connection {
         Ok(())
     }
 
-    pub async fn access_token(
+    pub async fn finalize(
+        &mut self,
+        store: Box<dyn OAuthStore + Sync + Send>,
+        code: &str,
+    ) -> Result<()> {
+        // If we finalized while waiting for the lock return early.
+        match self.status {
+            // Pending is the expected status for a new connection.
+            ConnectionStatus::Pending => (),
+            // We allow calling finalized twice with the same code (user messes up or refresh).
+            // Otherwise we error.
+            ConnectionStatus::Finalized => match self.unseal_authorization_code()? {
+                Some(c) => {
+                    // If it's finalized and the code matches, we return early.
+                    if c == code {
+                        return Ok(());
+                    }
+                    Err(anyhow::anyhow!("Connection is already finalized"))?
+                }
+                // If we have a finalized connection without `authorization_code` we 500.
+                None => unreachable!(),
```

Comment:
Can we add an error message to help debugging the day we reaches this code?

Reply:
(done)